font shorthand should not reject values that happen to have CSS-wide keywords as non-first identifiers in a font family name

darinadler opened the pull request:
#### 3dde8827051a9b051f6614f5956ee9ca6e1cbc31
<pre>
font shorthand should not reject values that happen to have CSS-wide keywords as non-first identifiers in a font family name
<a href="https://bugs.webkit.org/show_bug.cgi?id=246570">https://bugs.webkit.org/show_bug.cgi?id=246570</a>
rdar://problem/101459948

Reviewed by Tim Nguyen.

* LayoutTests/fast/css/font-shorthand-mix-inherit-expected.txt: Updated to match the new test below
including different expectations for some test cases.
* LayoutTests/fast/css/font-shorthand-mix-inherit.html: Changed to expect allowing CSS-wide keywords
in cases where they are supposed to be allowed. Refined the test to be simpler and so the test output
is easier to understand.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_family_parsing-expected.txt:

* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/target-text-007-expected.txt: Removed.
This was an unnecessary ignored file since this is a reference test. Not related to the code change.

* Source/WebCore/css/FontSelectionValueInlines.h: Removed unused isCSS21Weight function.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantCSS21): Deleted. Rolled this in to the one place it was used. This
helps make it clear that place was doing some unnecessary redundant work, and gets rid of the fairly
unclear name of this function.
(WebCore::CSSPropertyParser::consumeFont): Removed unnecessary and incorrect check for inherit and
initial keywords. The individual parsing steps already forbid those as appropriate, and allow them
in the case that was causing this bug. I have some doubts about whether this function correctly
returns true vs. false in all cases and what it consumes when it fails, but those don&apos;t seem to have
observable effects at this time, so I left that as is. Also tightened up the idiom in the loop that
parses the style, variant, weight, and stretch so it&apos;s easier to see an overview of it.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp: Fixed the &quot;interger&quot; typo throughout
this file.
(WebCore::CSSPropertyParserHelpers::consumeIntegerTypeRaw):
(WebCore::CSSPropertyParserHelpers::consumeNumberRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumePercentRaw): Made local to this file, and removed
the ValueRange argument.
(WebCore::CSSPropertyParserHelpers::consumePercentRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeLengthRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAngleRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeAngleOrPercent): Removed the ValueRange, UnitlessQuirk,
UnitlessZeroQuirk, and CSSValuePool arguments.
(WebCore::CSSPropertyParserHelpers::consumeLengthOrPercentRaw): Made local to this file
and removed the ValueRange and UnitlessQuirk arguments.
(WebCore::CSSPropertyParserHelpers::consumeAngleOrNumberRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAngleOrNumberRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeNumberOrPercentRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeNumberOrNoneRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightNumberRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightNumber): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeIdentRangeRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeFontVariantCSS21Raw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightKeywordValueRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightRaw): Moved the font weight keyword list
in here instead of a separate function.
(WebCore::CSSPropertyParserHelpers::consumeGenericFamilyRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontFamilyRaw): Made local to this file. Moved the
generic family range here instead of a separate function.
(WebCore::CSSPropertyParserHelpers::consumeFontSizeRaw): Made local to this file. Removed the
UnitlessQuirk argument.
(WebCore::CSSPropertyParserHelpers::consumeLineHeightRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeFontRaw): Removed unnecessary and incorrect check for
inherit and initial keywords. The individual parsing steps already forbid those as appropriate.
Also tightened up the idiom in the loop that parses the style, variant, weight, and stretch so
it&apos;s easier to see an overview of it.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Removed NumberOrPercentRaw,
NumberOrNoneRaw, PercentOrNoneRaw, NumberOrPercentOrNoneRaw, AngleOrNumberRaw,
AngleOrNumberOrNoneRaw, consumeFontWeightNumberRaw, consumeFontWeightNumber,
consumeLengthRaw, consumePercentRaw, consumeLengthOrPercentRaw, consumeAngleRaw,
consumeIdentRangeRaw, consumeFontVariantCSS21Raw, consumeFontWeightKeywordValueRaw,
consumeGenericFamilyRaw, consumeFontFamilyRaw, consumeFontSizeRaw, and consumeLineHeightRaw,
since none are used outside CSSPropertyParserHelpers.cpp.

Canonical link: <a href="https://commits.webkit.org/255894@main">https://commits.webkit.org/255894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1f3cacc3c848d46d58c4727d3f4a7520f57cd92

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93911 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103544 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163882 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97904 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3119 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31339 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86234 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99569 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99577 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2231 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80335 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29249 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84157 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72203 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37732 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17684 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35598 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18947 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4067 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41515 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41411 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38177 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5373,46 +5368,23 @@ bool CSSPropertyParser::consumeSystemFont(bool important)
-            if (fontStretch)
-                continue;
-        }
+        if (!fontStyle && (fontStyle = consumeFontStyle(m_range, m_context.mode, CSSValuePool::singleton())))
+            continue;
+        if (!fontVariantCaps && (fontVariantCaps = consumeIdent<CSSValueNormal, CSSValueSmallCaps>(m_range)))
+            continue;
+        if (!fontWeight && (fontWeight = consumeFontWeight(m_range)))
+            continue;
+        if (!fontStretch && (fontStretch = consumeFontStretchKeywordValue(m_range, CSSValuePool::singleton())))
+            continue;
```
Pre-existing bug, but this code actually breaks for cases like `normal normal normal 300 16px Arial` since it can't disambiguate `normal` if the non-normal values are not the order specified by the loop. Whereas `normal normal 300 normal 16px Arial` works. (This stood out to me because we had a similar bug with `list-style`, see https://bugs.webkit.org/show_bug.cgi?id=243991)

Testcase: `data:text/html,<span style="font: normal normal normal 300 16px Arial">Hello</span>`
Other browsers render this properly.

darinadler replied:
I’m happy to fix that; I know exactly how, it is relatively easy. But should we use a different pull request for it? Or do you want me to add to this one? The fix is quite simple but writing thorough test cases could be a challenge.

nt1m replied:
Separate PR would probably be better.

darinadler committed:
font shorthand should not reject values that happen to have CSS-wide keywords as non-first identifiers in a font family name
https://bugs.webkit.org/show_bug.cgi?id=246570
rdar://problem/101459948

Reviewed by Tim Nguyen.

* LayoutTests/fast/css/font-shorthand-mix-inherit-expected.txt: Updated to match the new test below
including different expectations for some test cases.
* LayoutTests/fast/css/font-shorthand-mix-inherit.html: Changed to expect allowing CSS-wide keywords
in cases where they are supposed to be allowed. Refined the test to be simpler and so the test output
is easier to understand.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_family_parsing-expected.txt:

* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/target-text-007-expected.txt: Removed.
This was an unnecessary ignored file since this is a reference test. Not related to the code change.

* Source/WebCore/css/FontSelectionValueInlines.h: Removed unused isCSS21Weight function.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantCSS21): Deleted. Rolled this in to the one place it was used. This
helps make it clear that place was doing some unnecessary redundant work, and gets rid of the fairly
unclear name of this function.
(WebCore::CSSPropertyParser::consumeFont): Removed unnecessary and incorrect check for inherit and
initial keywords. The individual parsing steps already forbid those as appropriate, and allow them
in the case that was causing this bug. I have some doubts about whether this function correctly
returns true vs. false in all cases and what it consumes when it fails, but those don't seem to have
observable effects at this time, so I left that as is. Also tightened up the idiom in the loop that
parses the style, variant, weight, and stretch so it's easier to see an overview of it.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp: Fixed the "interger" typo throughout
this file.
(WebCore::CSSPropertyParserHelpers::consumeIntegerTypeRaw):
(WebCore::CSSPropertyParserHelpers::consumeNumberRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumePercentRaw): Made local to this file, and removed
the ValueRange argument.
(WebCore::CSSPropertyParserHelpers::consumePercentRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeLengthRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAngleRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeAngleOrPercent): Removed the ValueRange, UnitlessQuirk,
UnitlessZeroQuirk, and CSSValuePool arguments.
(WebCore::CSSPropertyParserHelpers::consumeLengthOrPercentRaw): Made local to this file
and removed the ValueRange and UnitlessQuirk arguments.
(WebCore::CSSPropertyParserHelpers::consumeAngleOrNumberRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAngleOrNumberRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeNumberOrPercentRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeNumberOrNoneRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightNumberRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightNumber): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeIdentRangeRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeFontVariantCSS21Raw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightKeywordValueRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontWeightRaw): Moved the font weight keyword list
in here instead of a separate function.
(WebCore::CSSPropertyParserHelpers::consumeGenericFamilyRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeFontFamilyRaw): Made local to this file. Moved the
generic family range here instead of a separate function.
(WebCore::CSSPropertyParserHelpers::consumeFontSizeRaw): Made local to this file. Removed the
UnitlessQuirk argument.
(WebCore::CSSPropertyParserHelpers::consumeLineHeightRaw): Made local to this file.
(WebCore::CSSPropertyParserHelpers::consumeFontRaw): Removed unnecessary and incorrect check for
inherit and initial keywords. The individual parsing steps already forbid those as appropriate.
Also tightened up the idiom in the loop that parses the style, variant, weight, and stretch so
it's easier to see an overview of it.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Removed NumberOrPercentRaw,
NumberOrNoneRaw, PercentOrNoneRaw, NumberOrPercentOrNoneRaw, AngleOrNumberRaw,
AngleOrNumberOrNoneRaw, consumeFontWeightNumberRaw, consumeFontWeightNumber,
consumeLengthRaw, consumePercentRaw, consumeLengthOrPercentRaw, consumeAngleRaw,
consumeIdentRangeRaw, consumeFontVariantCSS21Raw, consumeFontWeightKeywordValueRaw,
consumeGenericFamilyRaw, consumeFontFamilyRaw, consumeFontSizeRaw, and consumeLineHeightRaw,
since none are used outside CSSPropertyParserHelpers.cpp.

Canonical link: https://commits.webkit.org/255894@main
